user/src/com/google/gwt/user/client/impl/WindowImpl.java

```diff
@@ -43,7 +62,12 @@ public native void initWindowCloseHandler() /*-{
       try {
         ret = $entry(@com.google.gwt.user.client.Window::onClosing())();
       } finally {
-        oldRet = oldOnBeforeUnload && oldOnBeforeUnload(evt);
+        try {
+          oldRet = oldOnBeforeUnload && oldOnBeforeUnload(evt);
+        } catch (unhandled) {
+          // Pass to uncaught exception handler, if registered
+          $wnd.onerror(unhandled);
```

Comment:
I actually had `console.log()` in mind. Calling `onerror` probably makes more sense as the exception would have ended up in that error handler as well, if it exists. Although personally I don't like it that much because the code doesn't stop executing here. So it is more like a warning log.

However calling `onerror` is more involved. First you need to check if `onerror` exists to not cause an error yourself and then you need to check if `unhandled` is an `Error` or not. If it is, you can pass in all five parameters `onerror(msg, sourceFileUrl, line, col, errorObj)` taken from `Error` otherwise only `onerror(unhandled)` is possible.

If `$wnd.onerror` does not exist the code could still fallback to `console.log(unhandled)`.

Reply:
The other option is to do like reportUncaughtException does: `setTimeout(function() {throw unhandled;},0)`. This will re-throw after the current code completes, and will either reach a registered onerror handler, or the browser will log it.

My preferences, to keep it simpler, from most preferred to least:
 * revert to original, just let it throw instead of returning a value
 * throw in a setTimeout
 * just console.error()
 * test the type of the error, and either invokie onerror or console.error()

Reply:
> throw in a setTimeout

Sounds good, but could we throw the JS exception and not the Java one? (c.f. https://github.com/gwtproject/gwt-core/issues/20)

Reply:
Would re-throwing in a setTimeout execute given that this is beforeUnload and once beforeUnload has been completed the page will be shutdown by the browser? If that is somewhat brittle I would lean towards reverting the change. The developer then has to fix the old beforeUnload handler to make ours work.

Reply:
Reverted back to how this was originally done.